drivers/dma/dma_intel_adsp_gpdma.c

```diff
@@ -129,9 +129,15 @@ static inline void intel_adsp_gpdma_llp_read(const struct device *dev,
 {
 #ifdef CONFIG_DMA_INTEL_ADSP_GPDMA_HAS_LLP
 	const struct intel_adsp_gpdma_cfg *const dev_cfg = dev->config;
+	uint32_t tmp;
 
-	*llp_l = dw_read(dev_cfg->shim, GPDMA_CHLLPL(channel));
+	tmp = dw_read(dev_cfg->shim, GPDMA_CHLLPL(channel));
 	*llp_u = dw_read(dev_cfg->shim, GPDMA_CHLLPU(channel));
+	*llp_l = dw_read(dev_cfg->shim, GPDMA_CHLLPL(channel));
+	if (tmp > *llp_l) {
+		/* re-read the LLPU value, as LLPL just wrapped */
+		*llp_u = dw_read(dev_cfg->shim, GPDMA_CHLLPU(channel));
+	}
```

Comment:
is this always called with interrupts blocked or can this be preempted? If the latter, at least in theory it could wrap again between lines 136 and 139? I think a traditional way to do this is to perform 3 measurements and *after* that to check for consistency. Potentially in a loop. Something line
```
for (l2 = read_low(), l1 = l2 + 1; l2 < l1; l2 = read_low()) {
    h = read_high();
    l1 = l2;
} 
```
or
```
for (h2 = read_high(), h1 = h2 + 1; h1 != h2; h2 = read_high()) {
    l = read_low();
    h1 = h2;
}
```

Reply:
Is it reasonable to think that the DMA moves 4294967295 bytes just like that?

Reply:
of course not, "this error can never happen" (TM) ;-) But that's the reason why this isn't a blocker, but just a comment. Also it "this error is purely theoretical" but it costs me almost nothing to account for it, I might as well decide to do so

Reply:
@ujfalusi can this be interrupted? If this is running in a simple thread context, it can be preempted for a while

Reply:
I think it can but likely for shorter period than 1.53 hours (96KHz, S32_LE, stereo)